static/compiler-picker.js

```diff
@@ -0,0 +1,188 @@
+        items: compilerId ? [compilerId] : [],
+        dropdownParent: 'body',
+        closeAfterSelect: true,
+        plugins: ['dropdown_input'],
+        maxOptions: 1000,
+        onChange: _.bind(function (val) {
+            if (val) {
+                ga.proxy('send', {
+                    hitType: 'event',
+                    eventCategory: 'SelectCompiler',
+                    eventAction: val,
+                });
+                this.onCompilerChange(val);
+            }
+        }, this),
+        duplicates: true,
+        render: {
+            option: function (data, escape) {
+                var extraClasses = data.$groups.indexOf(favoriteGroupName) !== -1 ? ' fav' : '';
+                return '<div class="d-flex' + extraClasses + '"><div>' + escape(data.name) + '</div>' +
+                    '<div title="Click to mark or unmark as a favorite" class="ml-auto toggle-fav">' +
+                    '<i class="fas fa-star"></i>' +
+                    '</div>' +
+                    '</div>';
+            },
+        },
+    });
+
+    $(this.tomSelect.dropdown_content).on('click', '.toggle-fav', _.bind(function (evt) {
+        evt.preventDefault();
+        evt.stopPropagation();
+
+        var optionElement = evt.currentTarget.closest('.option');
+        var clickedGroup = optionElement.parentElement.dataset.group;
+        var value = optionElement.dataset.value;
+        var data = this.tomSelect.options[value];
+        var isAddingNewFavorite = data.$groups.indexOf(favoriteGroupName) === -1;
+        var elemTop = optionElement.offsetTop;
+
+        if (isAddingNewFavorite) {
+            data.$groups.push(favoriteGroupName);
+            this.addToFavorites(data.id);
+        } else {
+            data.$groups.splice(data.group.indexOf(favoriteGroupName), 1);
+            this.removeFromFavorites(data.id);
+        }
+
+        this.tomSelect.updateOption(value, data);
+        this.tomSelect.refreshOptions(false);
+
+        if (clickedGroup !== favoriteGroupName) {
+            // If the user clicked on an option that wasn't in the top "Favorite" group, then we just added
+            // or removed a bunch of controls way up in the list. Find the new element top and adjust the scroll
+            // so the element that was just clicked is back under the mouse.
+            optionElement = this.tomSelect.getOption(value);
+            this.tomSelect.dropdown_content.scrollTop += (optionElement.offsetTop - elemTop);
+        }
+    }, this));
+};
+
+CompilerPicker.prototype.getOptions = function (langId, compilerId) {
+    var favorites = this.getFavorites();
+    return _.chain(this.compilerService.getCompilersForLang(langId))
+        .filter(function (e) {
+            return !e.hidden || e.id === compilerId;
+        })
+        .map(function (e) {
+            e.$groups = [e.group];
+            if (favorites[e.id])
+                e.$groups.push(favoriteGroupName);
```

Comment:
When the explorer page is reloaded, there's nothing in the data of the selected compiler to indicate which option in the dropdown should be selected: the option under "favorites" or the normal option.

```e.$groups.push(favoriteGroupName);``` will select the normal option

```e.$groups.unshift(favoriteGroupName);``` will select the option under "favorites"

Reply:
I _think_ the favoriteGroupName one is probably the best option. I'll make the change and see how I like it. Thanks!